client/app/services/dashboard.js

```diff
@@ -1,12 +1,49 @@
-import * as _ from 'underscore';
+import _ from 'underscore';
+
+function prepareWidgetsForDashboard(widgets) {
+  // Default height for auto-height widgets.
+  // Compute biggest widget size and choose between it and some magic number.
+  // This value should be big enough so auto-height widgets will not overlap other ones.
+  const defaultWidgetSizeY = Math.max(
+    _.chain(widgets)
+      .map(w => w.options.position.sizeY)
+      .max()
+      .value(),
+    20,
+  ) + 5;
+
+  // Fix layout:
+  // 1. sort and group widgets by row
+  // 2. update position of widgets in each row - place it right below
+  //    biggest widget from previous row
+  _.chain(widgets)
+    .sortBy(widget => widget.options.position.row)
+    .groupBy(widget => widget.options.position.row)
+    .reduce((row, widgetsAtRow) => {
+      let height = 1;
+      _.each(widgetsAtRow, (widget) => {
+        height = Math.max(
+          height,
+          widget.options.position.autoHeight
+            ? defaultWidgetSizeY
+            : widget.options.position.sizeY,
+        );
+        widget.options.position.row = row;
```

Comment:
Won't `position.row` will already be `row`?

I don't see how `height` or `row + height` is being later used?

Reply:
This function is used to ensure that widgets are sorted by rows/columns and does not overlap - because otherwise they may be placed in unpredictable way. The idea is to compute new positions for all widgets so they will have some space between - `gridstack` will then "flow" them up and all widgets will be placed at proper positions. Algorithm:

0. we have widgets at arbitrary positions, possibly overlapping;
1. groups all widgets by rows and orders groups by row number ascending - now we have logical rows, each row has a list of widgets at that row, but rows may still overlap;
2. Assign new row number for widgets at each row: start from `0`, and for each row use previous row number plus height of highest widget of previous row (for auto-height widgets use value computed before the loop - it's height of highest widget among all, or 20 - just a large enough value). At this step we ensured that rows does not overlap. Yes, most likely here will be large space between rows, but it does not matter - `gridstack` has no limitation for grid height (in opposite to `angular-gridster`).
3. Finally sort widgets by rows (at this step we have new row numbers) ascending, and by column (also ascending) - his will ensure that widgets will be added to grid from top top bottom and from left to right - `gridstack` tries to reflow widgets after each added, so if we add, for example, bottom-most widget first, it will "flow" up, and most top-left widget will be placed below it.

Reply:
I got confused by the `_.chain` syntax, but it's clear now.

Will the user see this initial reflow? Or gridstack takes care of it before rendering?

Reply:
No, no initial reflow